airbyte-integrations/connectors/source-mixpanel/source_mixpanel/source.py

```diff
@@ -0,0 +1,839 @@
+
+
+class Cohorts(MixpanelStream):
+    """Returns all of the cohorts in a given project.
+    API Docs: https://developer.mixpanel.com/reference/cohorts
+    Endpoint: https://mixpanel.com/api/2.0/cohorts/list
+
+    [{
+        "count": 150
+        "is_visible": 1
+        "description": "This cohort is visible, has an id = 1000, and currently has 150 users."
+        "created": "2019-03-19 23:49:51"
+        "project_id": 1
+        "id": 1000
+        "name": "Cohort One"
+    },
+    {
+        "count": 25
+        "is_visible": 0
+        "description": "This cohort isn't visible, has an id = 2000, and currently has 25 users."
+        "created": "2019-04-02 23:22:01"
+        "project_id": 1
+        "id": 2000
+        "name": "Cohort Two"
+    }
+    ]
+
+    """
+
+    data_field = None
+    primary_key = "id"
+
+    def path(self, **kwargs) -> str:
+        return "cohorts/list"
+
+
+class FunnelsList(MixpanelStream):
+    """List all funnels
+    API Docs: https://developer.mixpanel.com/reference/funnels#funnels-list-saved
+    Endpoint: https://mixpanel.com/api/2.0/funnels/list
+    """
+
+    primary_key = "funnel_id"
+    data_field = None
+
+    def path(self, **kwargs) -> str:
+        return "funnels/list"
+
+
+class DateSlicesMixin:
+    def stream_slices(
+        self, sync_mode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        date_slices = []
+
+        # use the latest date between self.start_date and stream_state
+        start_date = self.start_date
+        if stream_state:
+            # Remove time part from state because API accept 'from_date' param in date format only ('YYYY-MM-DD')
+            # It also means that sync returns duplicated entries for the date from the state (date range is inclusive)
+            stream_state_date = datetime.fromisoformat(stream_state["date"]).date()
+            start_date = max(start_date, stream_state_date)
+
+        # use the lowest date between start_date and self.end_date, otherwise API fails if start_date is in future
+        start_date = min(start_date, self.end_date)
+
+        # move start_date back <attribution_window> days to sync data since that time as well
+        start_date = start_date - timedelta(days=self.attribution_window)
+
+        while start_date <= self.end_date:
+            end_date = start_date + timedelta(days=self.date_window_size - 1)  # -1 is needed because dates are inclusive
+            date_slices.append(
+                {
+                    "start_date": str(start_date),
+                    "end_date": str(min(end_date, self.end_date)),
+                }
+            )
+            # add 1 additional day because date range is inclusive
+            start_date = end_date + timedelta(days=1)
+
+        # reset rate_limit if we expect less requests (1 req per stream) than it is allowed by API rate_limit
```

Comment:
See my previous comment about rate limits and exponential backoff. I think we would avoid the need to manually up the rate limit here (which is still brittle imo) by having an exponential backoff since that could start fast and only slow down if required

Reply:
Planned API limits:
Export stream - 60 reqs/hour (1 req for 60 sec).
Another streams - 400 reqs/hour (1 req for 9 sec)

When I used our default exponential backoff approach I got 'too many requests' error a few times.

So that's why I decided to run requests with:

- 1 sec delay if we know that the total number of requests is going to be less than API rate limit.
- 'planned' secs - delay if there are too many planned requests and we need to overcome the error.


I am still new in connector development so I really need your advice here.

Could you please confirm that it is better to remove this suggested approach and just use the default exponential backoff approach?

Thank you!

Reply:
exponential backoff is more robust imo. 
'too many requests' errors are expected since this is how the backoff is triggered each time. 
My main thinking here is with this current implementation, what happens in the case that the user uses the same API key elsewhere and uses up half of the rate allowed. Will this connector then error because it expects to be allowed the full rate or does it handle that gracefully?

Reply:
We have exponential [backoff today in the CDK](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/http/rate_limiting.py#L41) - if a request fails this automatically triggers.

I'm more concerned with thundering herd problems if there are multiple threads kicking off the request. Since Python has a global thread I don't think this will happen, though I might be wrong.

Although I do agree with ya George that this isn't the best, I'm inclined to say this is 'okay' for now. We should definitely create a follow up ticket to address this so the work isn't forgotten. Does that sound sane?

This is also making me think we need a mechanism to set QPS in the CDK. We can start off with something simple and later integrate the backoff function. WDYT @sherifnada ?

Reply:
Ah nice, didn't realise that!

Unless we're explicitly making async requests then they'll be global thread so shouldn't see thundering herd issues.

Yeah I can agree with good enough given that this will have backoff retries from cdk. My concern was that we'd fail as soon as we got a bad response but that's now moot

Reply:
Great! @midavadim can you please bring this up with @yaroslav-hrytsaienko and @sherifnada and follow up here with a ticket?

Reply:
https://github.com/airbytehq/airbyte/pull/4786

Here is an existing issue regarding the question we discussed here.